mlflow/pytorch/_pytorch_autolog.py

```diff
@@ -55,11 +52,207 @@ def _get_optimizer_name(optimizer):
+
+        :param trainer: pytorch lightning trainer instance
+        :param pl_module: pytorch lightning base module
+        """
+        self.client.set_tags(self.run_id, {"Mode": "training"})
+
+        params = {"epochs": trainer.max_epochs}
+
+        # TODO For logging optimizer params - Following scenarios are to revisited.
+        # 1. In the current scenario, only the first optimizer details are logged.
+        #    Code to be enhanced to log params when multiple optimizers are used.
+        # 2. mlflow.log_params is used to store optimizer default values into mlflow.
+        #    The keys in default dictionary are too short, Ex: (lr - learning_rate).
+        #    Efficient mapping technique needs to be introduced
+        #    to rename the optimizer parameters based on keys in default dictionary.
+
+        if hasattr(trainer, "optimizers"):
+            optimizer = trainer.optimizers[0]
+            params["optimizer_name"] = _get_optimizer_name(optimizer)
+
+            if hasattr(optimizer, "defaults"):
+                params.update(optimizer.defaults)
+
+        self.client.log_params(self.run_id, params)
+        self.client.flush(synchronous=True)
+
+    @rank_zero_only
+    def on_train_end(self, trainer, pl_module):
+        """
+        Logs the model checkpoint into mlflow - models folder on the training end
+
+        :param trainer: pytorch lightning trainer instance
+        :param pl_module: pytorch lightning base module
+        """
+        # manually flush any remaining metadata from training
+        self.metrics_logger.flush()
+        self.client.flush(synchronous=True)
+
+    @rank_zero_only
+    def on_test_end(self, trainer, pl_module):
+        """
+        Logs accuracy and other relevant metrics on the testing end
+
+        :param trainer: pytorch lightning trainer instance
+        :param pl_module: pytorch lightning base module
+        """
+        self.client.set_tags(self.run_id, {"Mode": "testing"})
+        self.client.flush(synchronous=True)
+
+        self.metrics_logger.record_metrics(
+            {key: float(value) for key, value in trainer.callback_metrics.items()}
+        )
+        self.metrics_logger.flush()
+
+
+def _log_early_stop_params(early_stop_callback, client, run_id):
+    """
+    Logs Early Stop parameters into mlflow
+
+    :param early_stop_callback: Early stopping callback dict
+    """
+    params = {}
+
+    if hasattr(early_stop_callback, "monitor"):
+        params["monitor"] = early_stop_callback.monitor
+
+    if hasattr(early_stop_callback, "mode"):
+        params["mode"] = early_stop_callback.mode
+
+    if hasattr(early_stop_callback, "patience"):
+        params["patience"] = early_stop_callback.patience
+
+    if hasattr(early_stop_callback, "min_delta"):
+        params["min_delta"] = early_stop_callback.min_delta
+
+    if hasattr(early_stop_callback, "stopped_epoch"):
+        params["stopped_epoch"] = early_stop_callback.stopped_epoch
```

Comment:
Can we simplify this code here?
```suggestion
    params = {
        p: getattr(early_stop_callback, p)
        for p in ["minitor", "mode", "patience", "min_delta", "stopped_epoch"]
        if hasattr(early_stop_callback, p)
    }
```

Reply:
I've also improved the docstring for this method and added a docstring to `_log_early_stop_metrics`